### ISSUE TITLE
Occasional 404/502 Errors from Proxy

### ISSUE THREAD

**Issue by chainzero** [akash-network/rpc-proxy#14]
When testing the current implementation/deployment of the RPC Proxy approx 8 out 10 requests are serviced properly but 10-20% of requests receive 404 not found or 502 forbidden responses.

Do not find any evidence in the service logs indicating source or nature of the issue.

Version - `0.0.1-rc2`

**Comment by caarlos0** [comment]
that seems to be running an older version of this project.. the most recent one should have the status in that log, and also properly ignore servers that are erroring 

see #12

**Comment by caarlos0** [comment]
yeah, this is not the latest version: https://peers.akash.network

**Comment by chainzero** [comment]
Installed just cut release `v0.0.1-rc3` and indeed status messages are written to logs in this version.

When request receives 404/502 response such as:

```
curl -ks https://peers.akash.network:443/rpc
<html>
<head><title>502 Bad Gateway</title></head>
<body>
<center><h1>502 Bad Gateway</h1></center>
<hr><center>nginx/1.26.1</center>
</body>
</html>
```

RPC proxy server logs show:

```
Aug 27 13:12:10 rpc-proxy akash-rpc-proxy[54476]: 2024/08/27 13:12:10 INFO request done name=Stakewolle avg=294.04119ms last=294.04119ms status=404
Aug 27 13:14:31 rpc-proxy akash-rpc-proxy[54476]: 2024/08/27 13:14:31 INFO request done name=Notional avg=398.930145ms last=398.930145ms status=502
```

**Comment by troian** [comment]
@caarlos0 are those `Stakewolle` and `Notional` that gives errors?

**Comment by troian** [comment]
@caarlos0 can you capture reply status and rule out nodes with 404/502 errors?

**Comment by chainzero** [comment]
We could definitely consider removing nodes from seeds if those nodes consistently error.

But in addition - do we currently attempt an additional seed if the first attempted is unresponsive/returns error?

**Comment by caarlos0** [comment]
> @caarlos0 can you capture reply status and rule out nodes with 404/502 errors?

it already does that... but after a while it will try again... its designed for intermittent errors, not exactly for things that always error...



> But in addition - do we currently attempt an additional seed if the first attempted is unresponsive/returns error?

we don't, easy enough to do though, can pr it later today

**Comment by caarlos0** [comment]
> @caarlos0 are those `Stakewolle` and `Notional` that gives errors?

- https://public.stakewolle.com/cosmos/akash/rpc always 404
- https://rpc-akash.cosmos-spaces.cloud always 502
- https://akash-rpc.kleomedes.network always 502
- https://rpc-akash-ia.cosmosia.notional.ventures always 502
- https://akash.declab.pro:26601 always times out

these seem to be all the problematic ones

**Comment by troian** [comment]
> But in addition - do we currently attempt an additional seed if the first attempted is unresponsive/returns error?
if original node return error pass it to the client, don't try another node for node

**Comment by chainzero** [comment]
> cool!
> 
> there are a few nodes that I think _always_ error - maybe we should remove them from the seed file as well?

@caarlos0 - was there intent to remove these unresponsive nodes?  Still getting sporadic occurrences of 404/502 and these same, identified nodes are used in such cases.  Example - this morning getting 502 errors when `https://rpc-akash.cosmos-spaces.cloud/` is used.

**Comment by caarlos0** [comment]
I can remove them yes, @troian wdyt?

**Comment by caarlos0** [comment]
> * https://akash-rpc.kleomedes.network always 502

this one seems to be back up though

**Comment by troian** [comment]
@caarlos0 lets do the following.
if node replies with anything than 200
- pass reply back to user
- mark node as unhealthy.
- run healthcheck query (for example query validator set) until it starts replying 200 within 10 consecutive attempts.
- return node back to active set if query check pass.
- if node becomes unhealthy 3 times within 30minutes then exclude is completely from set until `rpc-proxy` restarts

**Comment by caarlos0** [comment]
that sounds a bit overcomplicated... maybe something like this works better: https://github.com/akash-network/rpc-proxy/pull/11

for the cases here, in which the server is always 5xx/404, it should solve the issue I think 🤔

**Comment by troian** [comment]
sadly nature of RPC nodes on cosmos needs sophisticated health checks, historically the were pretty bad.
for example node can become simply due to being asked way to many requested and get back to normal shortly

**Comment by nick134-bit** [comment]
> > @caarlos0 are those `Stakewolle` and `Notional` that gives errors?
> 
> * https://public.stakewolle.com/cosmos/akash/rpc always 404
> * https://rpc-akash.cosmos-spaces.cloud always 502
> * https://akash-rpc.kleomedes.network always 502
> * https://rpc-akash-ia.cosmosia.notional.ventures always 502
> * https://akash.declab.pro:26601 always times out
> 
> these seem to be all the problematic ones

https://public.stakewolle.com/cosmos/akash/rpc/  <- note the last slash, is causing issues because they seem to use a proxy with the public available rpcs themself but dont seem to filter them. cosmos-spaces and notional are down consistently and handled. Other works for me.
